### ISSUE TITLE
Handling of duplicate transfer in committed state

### ISSUE THREAD

**Issue by sri-miriyala** [mojaloop/project#424]
When I was checking duplicate transfer handling in transfers-prepare (ML story#309), I sent a duplicate transfer with same parameters when the actual transfer is in committed state. I'm expecting a switch call back to payerfsp with a transfer success message from the actual transfer. Instead, I noticed switch is sending that transfer to payeefsp for processing and showing a committed state. Pls see below ml-adapter notification handler logs for reference.
![Screen Shot 2018-08-30 at 1.38.24 AM.png](https://images.zenhubusercontent.com/5ae0aa2d4b5806bc2bd09f9c/31b74fb5-88e7-45d8-aef8-67787ad15a55)

**Solution**
  - [x] Fix the central ledger code to send correct notification message
  - [x] Fix the ml-api-adapter to handle the new notification message 

**PRs**
- [x] https://github.com/mojaloop/central-ledger/pull/127
- [x] https://github.com/mojaloop/ml-api-adapter/pull/46

**Comment by mdebarros** [comment]
Note that I have created a new issue to update the documentation as follows: https://github.com/mojaloop/docs/issues/69

PRs created by @shashi165 
- https://github.com/mojaloop/central-ledger/pull/127
- https://github.com/mojaloop/ml-api-adapter/pull/46

**Comment by elnyry** [comment]
closing this out as both PRs are approved, merged

**Comment by sri-miriyala** [comment]
I reopened this issue as I did not complete my testing. I will close it once my testing is complete.

**Comment by sri-miriyala** [comment]
I can not test this feature until 3.5 fulfilment fix is redeployed.

**Comment by sri-miriyala** [comment]
Sam, if the fulfilment request from payeefsp rejected due to invalid fulfilment, where is this activity recorded in DB? I saw transfersStateChange table and I noticed just PREPARE and RESERVED status for that transfer request, but nothing about this activity.

**Comment by elnyry** [comment]
Hi Sri, if a fulfilment request from payeefsp has an invalid fulfilment, then its a validation failure, not a Rejection and so it does make sense why you just noticed PREPARE, RESERVED status in the transferStateChange table (It should be timed-out later, if no valid fulfilment comes in).

**Comment by sri-miriyala** [comment]
Sorry.... this discussion is not for this bug, but for #436. So moving there.

**Comment by sri-miriyala** [comment]
Tested and noticed the expected behavior in payerfsp. But I see an additional tag "extensionList" which is null, which I'm not sending in my transfers request. Need to discuss and then close this issue. Below is the response in payerfsp.

 In Payerfsp. TransferId: f31515ef-2159-4c4c-acdb-b34e44f640fc Body: {"transferId":"f31515ef-2159-4c4c-acdb-b34e44f640fc","amount":{"currency":"USD","amount":"110.00"},"transferState":"COMMITTED","completedTimestamp":"2018-09-17T05:34:35.000Z","ilpPacket":"AQAAAAAAAADIEHByaXZhdGUucGF5ZWVmc3CCAiB7InRyYW5zYWN0aW9uSWQiOiIyZGY3NzRlMi1mMWRiLTRmZjctYTQ5NS0yZGRkMzdhZjdjMmMiLCJxdW90ZUlkIjoiMDNhNjA1NTAtNmYyZi00NTU2LThlMDQtMDcwM2UzOWI4N2ZmIiwicGF5ZWUiOnsicGFydHlJZEluZm8iOnsicGFydHlJZFR5cGUiOiJNU0lTRE4iLCJwYXJ0eUlkZW50aWZpZXIiOiIyNzcxMzgwMzkxMyIsImZzcElkIjoicGF5ZWVmc3AifSwicGVyc29uYWxJbmZvIjp7ImNvbXBsZXhOYW1lIjp7fX19LCJwYXllciI6eyJwYXJ0eUlkSW5mbyI6eyJwYXJ0eUlkVHlwZSI6Ik1TSVNETiIsInBhcnR5SWRlbnRpZmllciI6IjI3NzEzODAzOTExIiwiZnNwSWQiOiJwYXllcmZzcCJ9LCJwZXJzb25hbEluZm8iOnsiY29tcGxleE5hbWUiOnt9fX0sImFtb3VudCI6eyJjdXJyZW5jeSI6IlVTRCIsImFtb3VudCI6IjIwMCJ9LCJ0cmFuc2FjdGlvblR5cGUiOnsic2NlbmFyaW8iOiJERVBPU0lUIiwic3ViU2NlbmFyaW8iOiJERVBPU0lUIiwiaW5pdGlhdG9yIjoiUEFZRVIiLCJpbml0aWF0b3JUeXBlIjoiQ09OU1VNRVIiLCJyZWZ1bmRJbmZvIjp7fX19","fulfilment":"XoSz1cL0tljJSCp_VtIYmPNw-zFUgGfbUqf69AagUzY","condition":"HOr22-H3AfTDHrSkPjJtVPRdKouuMkDXTR4ejlQa8Ks","expiration":"2018-09-20T10:50:26.000Z","extensionList":{"extension":[]}}

**Comment by elnyry** [comment]
Thanks Sri.

Is the presence of an empty extensionList an issue? Needs some investigation into why its being sent, its not a big issue however, since it still conforms to the Data Model and nothing wrong is being sent.

If it is an issue, please raise a bug and we can look into it.

**Comment by sri-miriyala** [comment]
Sam...I don't think sending empty tags is appropriate. Moreover if it is not coming in the incoming request, sending empty tags on the outbound response, atleast to me does not sound reasonable. But, if the team feels otherwise upon further discussion, I am OK with that too.

**Comment by elnyry** [comment]
By the way Sri, where are you getting the below message from - I mean is it from internal logs or is it from a callback/notification?

Body: {"transferId":"f31515ef-2159-4c4c-acdb-b34e44f640fc","amount":{"currency":"USD","amount":"110.00"},"transferState":"COMMITTED","completedTimestamp":"2018-09-17T05:34:35.000Z","ilpPacket":"AQAAAAAAAADIEHByaXZhdGUucGF5ZWVmc3CCAiB7InRyYW5zYWN0aW9uSWQiOiIyZGY3NzRlMi1mMWRiLTRmZjctYTQ5NS0yZGRkMzdhZjdjMmMiLCJxdW90ZUlkIjoiMDNhNjA1NTAtNmYyZi00NTU2LThlMDQtMDcwM2UzOWI4N2ZmIiwicGF5ZWUiOnsicGFydHlJZEluZm8iOnsicGFydHlJZFR5cGUiOiJNU0lTRE4iLCJwYXJ0eUlkZW50aWZpZXIiOiIyNzcxMzgwMzkxMyIsImZzcElkIjoicGF5ZWVmc3AifSwicGVyc29uYWxJbmZvIjp7ImNvbXBsZXhOYW1lIjp7fX19LCJwYXllciI6eyJwYXJ0eUlkSW5mbyI6eyJwYXJ0eUlkVHlwZSI6Ik1TSVNETiIsInBhcnR5SWRlbnRpZmllciI6IjI3NzEzODAzOTExIiwiZnNwSWQiOiJwYXllcmZzcCJ9LCJwZXJzb25hbEluZm8iOnsiY29tcGxleE5hbWUiOnt9fX0sImFtb3VudCI6eyJjdXJyZW5jeSI6IlVTRCIsImFtb3VudCI6IjIwMCJ9LCJ0cmFuc2FjdGlvblR5cGUiOnsic2NlbmFyaW8iOiJERVBPU0lUIiwic3ViU2NlbmFyaW8iOiJERVBPU0lUIiwiaW5pdGlhdG9yIjoiUEFZRVIiLCJpbml0aWF0b3JUeXBlIjoiQ09OU1VNRVIiLCJyZWZ1bmRJbmZvIjp7fX19","fulfilment":"XoSz1cL0tljJSCp_VtIYmPNw-zFUgGfbUqf69AagUzY","condition":"HOr22-H3AfTDHrSkPjJtVPRdKouuMkDXTR4ejlQa8Ks","expiration":"2018-09-20T10:50:26.000Z","extensionList":{"extension":[]}}

**Comment by sri-miriyala** [comment]
@elnyry I am seeing this in the callback. Did not check the logs.

**Comment by sri-miriyala** [comment]
This has been tested and met functionlaity. Hence closing this issue.  I will open a bug for the additional extensionList tag.
